instrumentation/servlet/servlet-jakarta/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/service/JakartaServletServiceAdvice.java

```diff
@@ -0,0 +1,132 @@
+import io.opentelemetry.javaagent.instrumentation.api.CallDepthThreadLocalMap;
+import io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.ServletResponse;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.util.concurrent.atomic.AtomicBoolean;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+
+public class JakartaServletServiceAdvice {
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static void onEnter(
+      @Advice.This(typing = Assigner.Typing.DYNAMIC) Object servletOrFilter,
+      @Advice.Argument(value = 0, readOnly = false) ServletRequest request,
+      @Advice.Argument(value = 1, readOnly = false) ServletResponse response,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    CallDepthThreadLocalMap.incrementCallDepth(AppServerBridge.getCallDepthKey());
+    if (!(request instanceof HttpServletRequest) || !(response instanceof HttpServletResponse)) {
+      return;
+    }
+
+    HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+
+    Context attachedContext = tracer().getServerContext(httpServletRequest);
+    if (attachedContext != null) {
+      // We are inside nested servlet/filter/app-server span, don't create new span
+      if (tracer().needsRescoping(attachedContext)) {
+        attachedContext =
+            tracer().updateContext(attachedContext, servletOrFilter, httpServletRequest);
+        scope = attachedContext.makeCurrent();
+        return;
+      }
+
+      // We already have attached context to request but this could have been done by app server
+      // instrumentation, if needed update span with info from current request.
+      Context currentContext = Java8BytecodeBridge.currentContext();
+      Context updatedContext =
+          tracer().updateContext(currentContext, servletOrFilter, httpServletRequest);
+      if (updatedContext != currentContext) {
+        // runOnceUnderAppServer updated context, need to re-scope
+        scope = updatedContext.makeCurrent();
+      }
+      return;
+    }
+
+    Context currentContext = Java8BytecodeBridge.currentContext();
+    if (currentContext != null
+        && Java8BytecodeBridge.spanFromContext(currentContext).isRecording()) {
+      // We already have a span but it was not created by servlet instrumentation.
+      // In case it was created by app server integration we need to update it with info from
+      // current request.
+      Context updatedContext =
+          tracer().updateContext(currentContext, servletOrFilter, httpServletRequest);
+      if (currentContext != updatedContext) {
+        // updateContext updated context, need to re-scope
+        scope = updatedContext.makeCurrent();
+      }
+      return;
+    }
```

Comment:
The start part is quite messy because it modifies several local variables injected to the instrumented method. The end part could have some common part extracted, but only the part that does not deal with async, but I couldn't think of a good name for that extracted method.

Reply:
ya, makes sense